File: python_test_crate/tests/08_hsm.py
from typing import Any, Optional, List, Dict, Callable

class HSMForwardFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class HSMForwardCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'HSMForwardCompartment':
        c = HSMForwardCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class HSMForward:
    def __init__(self):
        self._state_stack = []
        self._return_value = None
        self.log =         []
        self.__compartment = HSMForwardCompartment("Child")
        self.__next_compartment = None
        __frame_event = HSMForwardFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = HSMForwardFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = HSMForwardFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = HSMForwardFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def event_a(self):
        __e = HSMForwardFrameEvent("event_a", None)
        self.__kernel(__e)

    def event_b(self):
        __e = HSMForwardFrameEvent("event_b", None)
        self.__kernel(__e)

    def get_log(self) -> list:
        self._return_value = None
        __e = HSMForwardFrameEvent("get_log", None)
        self.__kernel(__e)
        return self._return_value

    def _state_Child(self, __e):
        if __e._message == "event_a":
            self.log.append("Child:event_a")
        elif __e._message == "event_b":
            self.log.append("Child:event_b_forward")
            self._state_Parent(__e)
        elif __e._message == "get_log":
            self._return_value = self.log
            __e._return = self._return_value
            return

    def _state_Parent(self, __e):
        if __e._message == "event_a":
            self.log.append("Parent:event_a")
        elif __e._message == "event_b":
            self.log.append("Parent:event_b")
        elif __e._message == "get_log":
            self._return_value = self.log
            __e._return = self._return_value
            return


def main():
    print("=== Test 08: HSM Forward ===")
    s = HSMForward()

    # event_a should be handled by Child (no forward)
    s.event_a()
    log = s.get_log()
    assert "Child:event_a" in log, f"Expected 'Child:event_a' in log, got {log}"
    print(f"After event_a: {log}")

    # event_b should forward to Parent
    s.event_b()
    log = s.get_log()
    assert "Child:event_b_forward" in log, f"Expected 'Child:event_b_forward' in log, got {log}"
    assert "Parent:event_b" in log, f"Expected 'Parent:event_b' in log (forwarded), got {log}"
    print(f"After event_b (forwarded): {log}")

    print("PASS: HSM forward works correctly")

if __name__ == '__main__':
    main()

File: rust_test_crate/tests/08_hsm.rs
use std::collections::HashMap;

#[derive(Clone, Debug)]
struct HSMForwardFrameEvent {
    message: String,
}

impl HSMForwardFrameEvent {
    fn new(message: &str) -> Self {
        Self { message: message.to_string() }
    }
}

#[derive(Clone)]
struct HSMForwardCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<HSMForwardFrameEvent>,
}

impl HSMForwardCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum HSMForwardStateContext {
    Child,
    Parent,
    Empty,
}

impl Default for HSMForwardStateContext {
    fn default() -> Self {
        HSMForwardStateContext::Child
    }
}

pub struct HSMForward {
    _state_stack: Vec<(String, HSMForwardStateContext)>,
    __compartment: HSMForwardCompartment,
    __next_compartment: Option<HSMForwardCompartment>,
    log: Vec<String>,
}

impl HSMForward {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            log: Vec::new(),
            __compartment: HSMForwardCompartment::new("Child"),
            __next_compartment: None,
        };
let __frame_event = HSMForwardFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: HSMForwardFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = HSMForwardFrameEvent::new("$<");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = HSMForwardFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = HSMForwardFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &HSMForwardFrameEvent) {
match self.__compartment.state.as_str() {
    "Child" => self._state_Child(__e),
    "Parent" => self._state_Parent(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: HSMForwardCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Child" => HSMForwardStateContext::Child,
    "Parent" => HSMForwardStateContext::Parent,
    _ => HSMForwardStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = HSMForwardFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    HSMForwardStateContext::Child => {}
    HSMForwardStateContext::Parent => {}
    HSMForwardStateContext::Empty => {}
}
    }

    pub fn event_a(&mut self) {
let __e = HSMForwardFrameEvent::new("event_a");
self.__kernel(__e);
    }

    pub fn event_b(&mut self) {
let __e = HSMForwardFrameEvent::new("event_b");
self.__kernel(__e);
    }

    pub fn get_log(&mut self) -> Vec<String> {
let __e = HSMForwardFrameEvent::new("get_log");
match self.__compartment.state.as_str() {
            "Child" => self._s_Child_get_log(&__e),
            "Parent" => self._s_Parent_get_log(&__e),
            _ => Default::default(),
        }
    }

    fn _state_Child(&mut self, __e: &HSMForwardFrameEvent) {
match __e.message.as_str() {
    "event_a" => { self._s_Child_event_a(__e); }
    "event_b" => { self._s_Child_event_b(__e); }
    "get_log" => { self._s_Child_get_log(__e); }
    _ => {}
}
    }

    fn _state_Parent(&mut self, __e: &HSMForwardFrameEvent) {
match __e.message.as_str() {
    "event_a" => { self._s_Parent_event_a(__e); }
    "event_b" => { self._s_Parent_event_b(__e); }
    "get_log" => { self._s_Parent_get_log(__e); }
    _ => {}
}
    }

    fn _s_Child_event_b(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Child:event_b_forward".to_string());
self._state_Parent(__e);
    }

    fn _s_Child_get_log(&mut self, __e: &HSMForwardFrameEvent) -> Vec<String> {
self.log.clone()
    }

    fn _s_Child_event_a(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Child:event_a".to_string());
    }

    fn _s_Parent_event_b(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Parent:event_b".to_string());
    }

    fn _s_Parent_get_log(&mut self, __e: &HSMForwardFrameEvent) -> Vec<String> {
self.log.clone()
    }

    fn _s_Parent_event_a(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Parent:event_a".to_string());
    }
}


fn main() {
    println!("=== Test 08: HSM Forward ===");
    let mut s = HSMForward::new();

    // event_a should be handled by Child (no forward)
    s.event_a();
    let log = s.get_log();
    assert!(log.contains(&"Child:event_a".to_string()), "Expected 'Child:event_a' in log, got {:?}", log);
    println!("After event_a: {:?}", log);

    // event_b should forward to Parent
    s.event_b();
    let log = s.get_log();
    assert!(log.contains(&"Child:event_b_forward".to_string()), "Expected 'Child:event_b_forward' in log, got {:?}", log);
    assert!(log.contains(&"Parent:event_b".to_string()), "Expected 'Parent:event_b' in log (forwarded), got {:?}", log);
    println!("After event_b (forwarded): {:?}", log);

    println!("PASS: HSM forward works correctly");
}
